### PR TITLE
Extract plugin handling in a `PluginRegistry` class independent of the `Application`

### LINKED ISSUE #697: Refactor the plugins handling out of the `Application`
### Problem

The plugins are handled directly within the `Application`. This prevents the usage of plugins to initialize application attribute. The symptomatic case is the need to customize the service manager in JupyterLab.

Xref: 
- https://github.com/jupyterlab/jupyterlab/issues/15329
- https://github.com/jupyterlab/jupyterlab/pull/16141#issuecomment-2049120215

### Proposed Solution

Extracting the plugin resolver logic in its own class (because why should such great feat be hard linked to the Application). Once we get that the logic in index.js will be:

- Create the plugin resolver with all the plugins
- Provide that resolver to the lumino Application

> :warning: we will need to expose the resolver method in the Application to ensure backward compatibility

### PR DESCRIPTION
**Pull request by fcollonval**
Fixes #697 
Fixes https://github.com/jupyterlab/lumino/issues/586

Refactor the plugins handling outside of the `Application` class to be able to instantiate and use it independently.
Add a allowed/blocked list feature to the new class.

The new `PluginRegistry` class is defined in coreutils to avoid bringing the widgets package in the stack dependency of the plugin system consumer. That implies adding the algorithm package as dependency to coreutils.

### CODE REVIEW

File: packages/coreutils/src/plugins.ts [hunk 1]
```diff
@@ -0,0 +1,867 @@
...
+   * plugins activation and deactivation functions.
+   *
+   * It can only be set once.
+   *
+   * By default, it is `null`.
+   */
+  get application(): T {
+    return this._application;
+  }
+  set application(v: any) {
```
Comment (brichet) [inline packages/coreutils/src/plugins.ts:145]: Shouldn't we type the parameter here ?
```suggestion
  set application(v: T) {
```

File: packages/coreutils/src/typing.d.ts
Comment (brichet) [inline packages/coreutils/src/typing.d.ts:1]: Out of curiosity, why do we need this ?
Should it be defined in other packages using `console` ?

**Review by brichet — COMMENTED** [review]
Thanks @fcollonval, LGTM, at least everything work in a dev installation.

I just have a few comments.

File: packages/coreutils/src/typing.d.ts
Comment (fcollonval) [reply packages/coreutils/src/typing.d.ts:1]: This package is compatible for browser and nodejs environment.

The automatic way to make TypeScript happy is to mention in the tsconfig that the target is `DOM`. However node js does not implement all the DOM API. But it does implement the console. So this pleases TypeScript but prevent future change to reference for example `window` or `document` (not available in node js).

File: packages/coreutils/src/plugins.ts (hunk 1 above)
Comment (fcollonval) [reply packages/coreutils/src/plugins.ts:145]: Indeed it makes sense

**Comment by fcollonval** [comment]
Thanks for the review @brichet

**Review by brichet — APPROVED** [review]

**Pull request merged**